get_http.go

```diff
@@ -118,12 +118,55 @@ func (g *HttpGetter) GetFile(dst string, u *url.URL) error {
 		g.Client = httpClient
 	}
 
-	resp, err := g.Client.Get(u.String())
+	// check to see whether user has specified a range of bytes to download.
+	// if user has, but the range is invalid, fall back to downloading the
+	// whole file
+	byteRange, rangeErr := getByteRange(u)
+	if rangeErr != nil {
+		// log that we are downloading whole file even though user
+		// requested a byte range.
+		log.Printf("%s; going to disregard range request and download entire file",
+			rangeErr)
+	}
+
+	// Create new Request here because if it is a range request, we need to set
+	// Range headers on the request object.
+	req, err := http.NewRequest("HEAD", u.String(), nil)
+	if err != nil {
+		return err
+	}
+
+	isPartialDownload := false
+	if rangeErr == nil && byteRange != nil {
+		// We first make a HEAD request so we can check if the server supports
+		// range queries. If the server/URL doesn't support HEAD requests,
+		// we just fall back to GET.
+
+		resp, err := g.Client.Do(req)
+		if err == nil && resp.StatusCode >= 200 && resp.StatusCode < 300 {
+			// If the HEAD request succeeded, then attempt to set the range
+			// query if we can.
+			if resp.Header.Get("Accept-Ranges") == "bytes" {
+				req.Header.Set("Range", fmt.Sprintf("bytes=%s-%s",
+					byteRange[0], byteRange[1]))
+				isPartialDownload = true
+			}
+		} else {
+			log.Printf("HEAD request for Range failed; falling back to full file download: %s",
+				err)
+		}
+	}
+
+	// Set the request to GET now, and redo the query to download
+	req.Method = "GET"
```

Comment:
I'm not actually sure about the API promise if you reuse an `*http.Request` struct. I would probably just create a new request here to be safe. As an example of why I would worry about this, the request's `Body` gets automatically closed by the transport. To me that implies that an `*http.Request` is not meant to be reused for multiple requests, even though it might work in this case.